---
title: "001-bioinformatics_scripts"
author: "Rebecca Cheek"
date: "2/5/2020"
output: html_document
---


These data are comprised of 152, adult, male Island Scrub Jays (*Aphelocoma insularis*) from accross Santa Cruz Island, California. 
Sample size is:

Western Oak  = 39
Central Oak = 60
Eastern Oak = 12
Western Pines = 25
Central Pines = 7
Eastern Pines = 9

*001- Process Radtags* 
*Description*: De-multiplex raw reads from sequencer by individual barcodes
*Associated Data*: Raw reads, barcode files for the 2 libraires. Output: processed fastq.gz 


Be sure to edit the header of each script for your own system to run it as a job. Here is mine as an example:

#!/bin/bash

#SBATCH --job-name=process_radtags
#SBATCH --output=/home/cheek/scratch_dir/ISSJs-genomics/process_samples/process_radtags.out
#SBATCH --error=/home/cheek/scratch_dir/ISSJs-genomics/process_samples/process_radtags.err
#SBATCH --mail-user=Rebecca.G.Cheek@gmail.com
#SBATCH --mail-type=END
#SBATCH --mem=10G

Because some of the barcodes overlapp, the process_radtags function from Stacks (v.2.53) needs to be run twice to demultiplex the two libraires.


Run the script for the first 72 individuals in Library 1. 

Filename for the script: "01-process_radtags.sh"

```{sh}
#!/bin/bash

#SBATCH --job-name=process_radtags_Lib1
#SBATCH --output=/home/cheek/scratch_dir/ISSJs-genomics/process_samples/process_radtags.out
#SBATCH --error=/home/cheek/scratch_dir/ISSJs-genomics/process_samples/process_radtags.err
#SBATCH --mail-user=Rebecca.G.Cheek@gmail.com
#SBATCH --mail-type=END
#SBATCH --mem=10G

process_radtags \
-p /home/cheek/scratch_dir/ISSJs-genomics/raw/rawdata_Lib_1 \
-b /home/cheek/scratch_dir/ISSJs-genomics/names_change_LIB1 \
-o /home/cheek/scratch_dir/ISSJs-genomics/process_samples/ \
--inline_null \
-i gzfastq \
-y gzfastq -e 'sbfI' -c -q -r -E 'phred33'

```


Run the same code for the 80 individuals in Library 2, but need to specify that it is in a different directory 
Filename for the script: "01.5-process_radtags2.sh"

```{sh}


#!/bin/bash

#SBATCH --job-name=process_radtags_Lib2
#SBATCH --output=/home/cheek/scratch_dir/ISSJs-genomics/process_samples/process_radtags.out
#SBATCH --error=/home/cheek/scratch_dir/ISSJs-genomics/process_samples/process_radtags.err
#SBATCH --mail-user=Rebecca.G.Cheek@gmail.com
#SBATCH --mail-type=END
#SBATCH --mem=10G

process_radtags \
-p /home/cheek/scratch_dir/ISSJs-genomics/raw/rawdata_Lib_2 \
-b /home/cheek/scratch_dir/ISSJs-genomics/names_change_LIB2 \
-o /home/cheek/scratch_dir/ISSJs-genomics/process_samples/ \
--inline_null \
-i gzfastq \
-y gzfastq -e 'sbfI' -c -q -r -E 'phred33'


```


All demultiplexed files are now in the process_samples directory. 


Two individuals, F09-123 (Coches) & F11-023 (PZ Valley) seemed to have failed sequencing as they have extremely small filesizes. These were also the same individuals that were dropped by P. Salerno for her analysis so will leave them out by adding them to the "dropped_individuals" directory.


Our final sampling breakdown for bioinformatics is:

Western Oak  = 38
Central Oak = 59
Eastern Oak = 12
Western Pines = 25
Central Pines = 7
Eastern Pines = 9



*002-Align_BWA*
*Description*: BWA is a software package for mapping low-divergent sequences against a large reference genome, BWA-MEM, which is the latest, is generally recommended for high-quality queries as it is faster and more accurate and has better performance for 70-100bp Illumina reads. Use BWA-mem to align reads to FLSJ reference genome
*Associated Data*: processed reads. Output bam files


```{sh}
#!/bin/bash

#SBATCH --job-name=issj_align
#SBATCH --output=/home/cheek/scratch_dir/ISSJs-genomics/aligned_reads/issj_aligned.out
#SBATCH --error=/home/cheek/scratch_dir/ISSJs-genomics/aligned_reads/issj_aligned.err
#SBATCH --mail-user=Rebecca.G.Cheek@gmail.com
#SBATCH --mail-type=END
#SBATCH --ntasks=1
#SBATCH --cpus-per-task=6
#SBATCH --mem=10G

GENOME=/home/cheek/scratch_dir/ISSJs-genomics/reference_genome/final.assembly.fasta.gz
BAMOUT=/home/cheek/scratch_dir/ISSJs-genomics/aligned_reads/bam
REPORTOUT=/home/cheek/scratch_dir/ISSJs-genomics/aligned_reads/reports

for sample in `ls *.fq.gz | sed "s/......$//"`
do
    bwa mem -t 6 \
    $GENOME $sample*.fq.gz | \
    samtools view -bhS - | samtools sort -o $BAMOUT/$sample.bam
    samtools flagstat $BAMOUT/$sample.bam &> $REPORTOUT/$sample.stats
done

```

*003-ref_map* 
*Description*: Program will execute Stacks pipeline to genotype and call SNPs using the population-wide data per locus. Makes use of the population map to specify all the samples in an analysis and determine which groupings to use for calculating summary statistics. If no population map is supplied the populations program computes these statistics with all samples as a single group.
*Associated Data*: aligned bam files. Output stacks files


```{sh}
#!/bin/bash

#SBATCH --job-name=issj_ref_map
#SBATCH --output=/home/cheek/scratch_dir/ISSJs-genomics/ref_map_local/ref_map_local.out
#SBATCH --error=/home/cheek/scratch_dir/ISSJs-genomics/ref_map_local/ref_map_local.err
#SBATCH --mail-user=Rebecca.G.Cheek@gmail.com
#SBATCH --mail-type=END
#SBATCH --ntasks=1
#SBATCH --cpus-per-task=8
#SBATCH --mem=10G

ALIGNED=/home/cheek/scratch_dir/ISSJs-genomics/aligned_reads/bam
POPMAP=/home/cheek/scratch_dir/ISSJs-genomics/popmap_local
REFOUT=/home/cheek/scratch_dir/ISSJs-genomics/ref_map_local/ref_map_out


ref_map.pl -T 8 --popmap $POPMAP -o $REFOUT --samples $ALIGNED -X "gstacks:--rm-pcr-duplicates"


```